### ISSUE TITLE
Feature request: new flag CMS_NO_SIGNING_TIME in CMS_sign() et al. to suppress addition of signing-time attribute

### ISSUE THREAD

**Issue by pjuhasz** [openssl/openssl#15777]
As of now, `CMS_sign()` and related functions unconditionally add a signing-time signed attribute to the SignerInfo of a SignedData CMS structure. Most of the time this is correct and proper, however, this attribute is not required, and there is one specific case where it is even explicitly disallowed: PAdES signatures for PDFs.

I propose that a new flag `CMS_NO_SIGNING_TIME` be added to the api that, when passed to `CMS_sign()` (and `CMS_final()` and the few others that accept such a flag), suppresses the addition of this attribute.

Analysis and notes:
Required effort: small
As far as I could see the addition of the attribute happens in only one place in `CMS_SignerInfo_sign()`, so the effective part of the modification would be one `if` around the offending code block, plus the addition of the `#define` in the header. (The flag could be exposed as a command line option of the `openssl cms` utility, which would mean more work, but that could be done as a second, separate step).

Risk: low
This is not a breaking change, and without the flag all code using openssl would work as before.

Reward: high
Currently it is not possible to create a fully compliant PAdES signature because of this header. The proposed simple change would make it possible, at little cost.

**Comment by mattcaswell** [comment]
If you would like to submit a pull request for this feature, please feel free. It is unlikely to be merged until post 3.0 however and wouldn't be back ported to stable releases.

**Comment by kaduk** [comment]
>  Most of the time this is correct and proper, however, this attribute is not required, and there is one specific case where it is even explicitly disallowed: PAdES signatures for PDFs.

Could you give a reference to the specification that prohibits the inclusion of the CMS signing-time attribute?  A google search for "pades pdf signature" found https://www.etsi.org/deliver/etsi_ts/102700_102799/10277801/01.01.01_60/ts_10277801v010101p.pdf and I failed to find a clause that stuck out to me as imposing the stated requirement.

**Comment by pjuhasz** [comment]
> Could you give a reference to the specification that prohibits the inclusion of the CMS signing-time attribute?
Sure.

The relevant document is ETSI EN 319 142-1, which is the final version of what you've linked: https://www.etsi.org/deliver/etsi_en/319100_319199/31914201/01.01.01_60/en_31914201v010101p.pdf

See Table 1 in this document, in particular the line "SPO: signing-time attribute in CMS signature". It says "shall not be present" for all profiles.

I know of at least one official-looking validator ( https://signatures-conformance-checker.etsi.org/pub/index.php , free but requires a registration) that croaks on signatures that contain the signing-time attribute.

**Comment by kaduk** [comment]
Thank you!

**Comment by IonutCorbu** [comment]
Hello! Is this flag added? Because I'm trying to create a PAdES signature and I have exactly this problem. I tried to remove this field with CMS_NOATTR in the call of CMS_add1_signer, but this is not removed. And as a result, my signature includes this field and it is not recognized by DSS or Adobe as a BASELINE-B PAdES signature. I also tried to remove it with CMS_signed_delete_attr and it is not working. 
Thank you and have a nice day!

**Comment by IonutCorbu** [comment]
I tried to build an OpenSSL version with the changes that you proposed here [NO_SIGNING_TIME](https://github.com/openssl/openssl/pull/15783/commits/e1b00300a0438a9804c605b223d6891ebfd6924a), but even if I added that flag, the signingTime is still present in CMS structure. Do you know why is this happening? Here is the slice of code:
```
CMS_SignerInfo* cms_info = CMS_add1_signer(cms, x509_cert, priv_key, EVP_get_digestbyname(hash_type.c_str()), CMS_NOATTR|CMS_NOSMIMECAP|CMS_NO_SIGNING_TIME);

	if (!CMS_signed_add1_attr_by_NID(cms_info, NID_pkcs9_contentType, V_ASN1_OBJECT, OBJ_nid2obj(NID_pkcs7_data), -1)) {
		CMS_ContentInfo_free(cms);
	}

	if (!CMS_signed_add1_attr_by_NID(cms_info, NID_pkcs9_messageDigest, V_ASN1_OCTET_STRING, digest.data(), digest.size())) {
		CMS_ContentInfo_free(cms);
	}

	ESS_SIGNING_CERT_V2* signing_cert = OSSL_ESS_signing_cert_v2_new_init(EVP_get_digestbyname(this->hash_type.c_str()), this->x509_cert, NULL, -1);

	int len_sign_cert = i2d_ESS_SIGNING_CERT_V2(signing_cert, NULL);
	unsigned char* encoded_data = (unsigned char*)malloc(len_sign_cert * sizeof(unsigned char));
	unsigned char* p = encoded_data;
	i2d_ESS_SIGNING_CERT_V2(signing_cert, &p);
	ASN1_STRING* seq = ASN1_STRING_new();
	ASN1_STRING_set(seq, encoded_data, len_sign_cert);
	CMS_signed_add1_attr_by_NID(cms_info, NID_id_smime_aa_signingCertificateV2, V_ASN1_SEQUENCE, seq, -1);

	CMS_set_detached(cms, 1);


	CMS_SignerInfo_sign(cms_info);
```

![image](https://github.com/openssl/openssl/assets/85988309/82f6cb3a-2618-40f5-a96d-728996098ae9)

**Comment by IonutCorbu** [comment]
Thank you for the contribution! I will struggle to do it work in my local version.

**Comment by pjuhasz** [comment]
I checked my modifications against the latest openssl master, and they appear to compile. Then I tested my signer solution with this modified version and it appears to work (i.e. it produced a CMS without the signing time attribute).

Looking at my signer code, I use the flags `CMS_BINARY | CMS_PARTIAL | CMS_NOSMIMECAP | CMS_DETACHED | CMS_NO_SIGNING_TIME` with CMS_add1_signer, so I don't use CMS_NOATTR.  That also means that I don't have to add the content type and message digest attributes manually, openssl does it just fine. Then I add the ESS signing cert and optionally, the OCSP response, and in the end I call CMS_final to do the actual signing.

**Comment by IonutCorbu** [comment]
Thank you for the information! Most probably, I had a problem during the compilation because it worked when I tried on Linux afterwards using the rules from [HACKING.md](https://github.com/openssl/openssl/blob/master/HACKING.md). I did not find an equivalent in Windows.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev, barring further input
